base-containers/base/bin/_run_student_intern

```diff
@@ -88,44 +51,50 @@ for name, (spath, append) in system_files.items():
         except IOError:
             exit(254)
 
+logger.info("student container started and received initial command")
+
 # Start the process
-print("Chdir")
 os.chdir(start_cmd["working_dir"])
-print("Popen")
-print(fds)
-
 set_limits = lambda: set_limits_user(user)  # To know if the command should be executed as root or worker
 if start_cmd["command"] is None or start_cmd["command"] == "":  # Avoid to run an empty command
     start_cmd["command"] = "echo 'info: student container started with no command set' "
 
-# Run the command, which in the case of ssh, acts like a setup_script
-p = subprocess.Popen(shlex.split(start_cmd["command"]), preexec_fn=set_limits, stdin=fds[0], stdout=fds[1], stderr=fds[2])
-# Handle the signals
-print("Signal thread")
-thread = threading.Thread(target=lambda: handle_signals(p, client), daemon=True)
-thread.start()
-# Wait until the subprocess exit
-print("Waiting for process to end")
-retval = p.wait()
-
+# Run the student code in its own subprocess and handle inputs/outputs
+if only_dockers:
+    p = subprocess.Popen(shlex.split(start_cmd["command"]), preexec_fn=set_limits, stdin=fds[0], stdout=fds[1],
+                         stderr=fds[2])
+    signal_thread = threading.Thread(target=lambda: handle_signals(p, socket_unix), daemon=True)
+    signal_thread.start()
+    retval = p.wait()
+else:
+    p = subprocess.Popen(shlex.split(start_cmd["command"]), preexec_fn=set_limits, stdin=subprocess.PIPE,
+                         stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    write_lock = threading.Lock()  # Lock to avoid stdout_handler and stderr_handler to write on the stdout_stream at the same time
```

Comment:
They are not combined.
stdout is sent via message of type "stdout" and stderr is sent via message of type "stderr".
But the fact is both are sent on the same asyncio stream (communication between agent and student_container).
Thus, in a situation were the student code is printing both on stdout and stderr at the same time, the lock will avoid having both threads pushing on the stream at the same time.

Reply:
OK. Was this necessary to have then multiplexed in the same stream then ?

Reply:
I believe having messages with type "stdout" and messages with type "stderr" sent on the same stream is not a big issue.
Having 2 separated streams is maybe a performance upgrade maybe not (it would mean handle the 2 streams both on the agent side and on the grading_container side).
I suggest @GuillaumeDerval to give us his opinion on this.

Reply:
Basically this exact same kind of stream (asyncio stream) was already used for different types of messages (such as run_student to tell agent to start container, student_retval to tell it finished etc...). Here the principle is the same, we have different types of messages (type stdout and stderr) transfered on this stream so I believe it should not be an issue.